### PR TITLE
Add install docs for Fedora and document note on virtualenv 

### LINKED ISSUE #5438: Review and update developer docs
We should routinely review the development environment docs, here:

https://docs.securedrop.org/en/stable/development/setup_development.html

with an eye to: 
- are all instructions still accurate and up-to-date when setting up an environment from scratch;
- are the instructions clear and well-organized;
- do they work across different environments (Linux [different distros], Mac, Windows).

@joaedwar has offered to take a stab at this in the next SecureDrop sprint; exact scope TBD.

### PR DESCRIPTION
**Pull request by nebelsprachen**
## Status

Ready for review 

## Description of Changes

Fixes #5438 

Changes proposed in this pull request: Adds docs on installing in Fedora and other Linux distributions. Also, adds note on creating a virtulalenv in Python3.  

## Testing

How should the reviewer test this PR?
1. Open docs site
2. Navigate to *Setting Up the Development Environment* section in docs
3. Navigate to *Documentation Guidelines* section

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

### If you made changes to documentation:

- [x] Doc linting (`make docs-lint`) passed locally

### If you added or updated a code dependency:

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [ ] I would like someone else to do the diff review

### CODE REVIEW

File: docs/development/documentation_guidelines.rst
Comment (zenmonkeykstop) [inline docs/development/documentation_guidelines.rst]: This is definitely a practice to encourage, so I think you could explicitly make it part of the  docs setup process rather than relegating it to a note. Elsewhere in the [docs](https://docs.securedrop.org/en/stable/development/setup_development.html#setting-up-a-multi-machine-environment) we recommend using `virtualenvwrapper`, so it might make sense to have this be consistent with those instructions. Leveraging includes could make sense here.

File: docs/development/setup_development.rst
Comment (zenmonkeykstop) [inline docs/development/setup_development.rst:43]: For consistency's sake, if we're adding a Fedora section for the dev environment setup we should consider adding it in the staging environment as well. That does get more complicated as there's all the virtualbox and/or libvirt setup to consider, though.

File: docs/development/setup_development.rst
Comment (zenmonkeykstop) [inline docs/development/setup_development.rst]: Both this and the original Ubuntu section are missing explanatory text as to why these commands should be run - I mean, it's pretty obvious but a line like "first make sure that git is available on your system" wouldn't hurt imo.

**Review by zenmonkeykstop — CHANGES_REQUESTED** [review]

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst]: Nit: It makes no difference in the output, but for readability, I would recommend shortening the length of the underline that creates the headline formatting (the `-------------`) to match the length of the headline.

File: docs/development/setup_development.rst
Comment (eloquence) [reply docs/development/setup_development.rst:43]: In the spirit of incremental improvements, adding a note about Fedora to the Docker instructions is IMO a good first step.

File: docs/development/documentation_guidelines.rst
Comment (nebelsprachen) [reply docs/development/documentation_guidelines.rst]: I think the doc flow should be reorganized, with a section that calls out setting up the VM. I included the note as a place holder until that re-architecture is finalized. Do you think this note is okay for the time being, or should I hold off until we restructure the flow?

Comment (eloquence) [reply docs/development/documentation_guidelines.rst]: You can follow the examples in `docs/includes` (`grep` through the docs to see how these are used) to create a re-usable text block `virtualenv.txt` which we can initially insert at this position and then place in other relevant sections of the text. 

Per discussion at standup, I suggest we go with `python3 -m venv .venv` as the canonical method (since it will always work without installing additional dependencies) and mention the existence of other convenience tools like `virtualenvwrapper`, which provides the `mkvirtualenv` shell function.

Comment (eloquence) [reply docs/development/documentation_guidelines.rst]: Also, as just discussed, let's add instructions for how to activate a virtualenv.

File: docs/development/setup_development.rst
Comment (eloquence) [reply docs/development/setup_development.rst]: I concur with @zenmonkeykstop that adding a bit more context here would be helpful, @joaedwar.

File: docs/includes/virtualenv.txt
Comment (eloquence) [inline docs/includes/virtualenv.txt]: In restructured text, double backticks (``) are used for inline code formatting, while single backticks italicize. All command references here should be enclosed in double backticks.

Comment (eloquence) [reply docs/includes/virtualenv.txt]: Typo: "virtual environmen" -> "virtual environment"

Comment (eloquence) [reply docs/includes/virtualenv.txt]: I would clarify the sentence ending with "using the program `pip`". a little bit. You can use `pip` to install packages with or without a virtualenv, what a virtualenv provides is a way to isolate the installation of specific dependencies of a specific application, so that the dependencies of another application (or updates to the OS at a whole) do not get in the way of that application's requirements

Comment (eloquence) [reply docs/includes/virtualenv.txt]: Typo in the command: `source venv/bin/activate` -> `source .venv/bin/activate` (note leading `.`)

Comment (eloquence) [reply docs/includes/virtualenv.txt]: Nit: "create and manage your virtual environments" -> "create and manage your virtual environment" (in this case you're just creating one, not multiple)

File: docs/development/documentation_guidelines.rst
Comment (eloquence) [inline docs/development/documentation_guidelines.rst:28]: The include directive breaks up the automatic numbering via the `#.` syntax (the 3rd item is again restarted at 1). You can fix that by just manually numbering the steps here, i.e. replacing `#.` with `1.`, `2.`, `3.`. You can verify the result of such changes by using the `make docs` autobuild and visiting http://localhost:8000/

**Comment by eloquence** [comment]
Thanks @joaedwar, this is looking great. Left a few comments, don't hesitate to ping if I can help further. Once we merge this PR, I suggest we consolidate the virtualenv instructions you're adding here with the virtualenv-related instructions in the setup chapter: https://docs.securedrop.org/en/stable/development/setup_development.html

File: docs/development/setup_development.rst
Comment (eloquence) [reply docs/development/setup_development.rst]: Noting a couple of things here on closer inspection:
1) `dnf update` is deprecated in favor of `dnf upgrade` which has the same meaning, see:
https://dnf.readthedocs.io/en/latest/cli_vs_yum.html#update-and-upgrade-commands-are-the-same

2) The `dnf update` / `dnf upgrade` commands behave _differently_ than `apt-get update`. These `dnf` commands attempt to update all packages installed on the system, while `apt-get update` only updates the package index.

To make the instructions equivalent to the Debian instructions, remove the `dnf update` command. Use only the `dnf install` command, which implicitly performs a metadata check to ensure it installs an up-to-date version.

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst:43]: Further down we are providing instructions for installing Docker under Ubuntu/Debian (the previous section), but they are now placed in the Fedora section.

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst]: Note that Docker also provides instructions for installing Docker CE on Fedora, which we should link to in the appropriate section (see earlier comment about the current section structure).

https://docs.docker.com/engine/install/fedora/

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst]: I would shorten to:
"Run the following commands to update the package index and to install Git and `make`"

Technically they are not Git repositories but distribution repositories, but we don't have to spell that out either way.

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst]: Same edit as above in the Debian section. As noted in another comment, just running `sudo dnf install git` will implicitly update the package index, while `sudo dnf update` updates all installed packages and is a deprecated alias for `sudo dnf upgrade`.

File: docs/includes/virtualenv.txt
Comment (eloquence) [inline docs/includes/virtualenv.txt]: Nit: Stray whitespace between "Python" and "packages". Regarding this text:

> You can install, upgrade, and remove packages using the program ``pip`` in an isolated environment. 

I think the goal of this paragraph is to provide a reusable explanation of how package installation works for people who are new to Python development. With this goal in mind, I would recommend wording like this:

> You can install, upgrade, and remove Python packages using `pip`, the Python package installer. By default, `pip` will attempt to install packages system-wide. Use `venv` or `virtualenv` to avoid ...

So that we:
1) introduce `pip` (what is its job in the Python world?)
2) explain its default behavior
3) show the value of virtual environments when using `pip` to manage dependencies.

**Comment by eloquence** [comment]
Getting really close. Requesting a few final tweaks :). Thanks for all your work on this so far @joaedwar.

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst]: Nit: commands->command

(It's just one command on Fedora)

File: docs/development/setup_development.rst
Comment (eloquence) [inline docs/development/setup_development.rst]: Right now this feels repetitive of the instructions that follow, and only partially (why mention Git and not `make` or Docker itself?). I would suggest we speak to the user's needs at a higher level here, e.g., language like the following (please feel free to reword as you see fit):

> Follow the instructions below to install the requirements for the Docker-based environment for your operating system.

**Comment by eloquence** [comment]
Only two comments left from my end and then we can kick it over to a maintainer for a final pass, thanks for your patience. :) You can ignore the CI failure, that's a result of #5470 which is currently being worked on.

Now would be a good time to squash the 15 commits into 1 if you're comfortable doing so; I'm also happy to go over that process in our 1:1.

**Comment by eloquence** [comment]
Update: #5470 has just been fixed, so if you rebase this PR on latest `develop` as part of squashing it, CI should go to green again.

**Comment by eloquence** [comment]
Closing in favor of #5478.

**Pull request closed without merging**
